modules/broker-core/src/main/java/org/wso2/broker/core/BindingsRegistry.java

```diff
@@ -32,65 +34,86 @@
  */
 public final class BindingsRegistry {
 
-    private final Map<String, BindingSet> routingKeyToBindingMap;
+    private final Map<String, BindingSet> bindingPatternToBindingsMap;
 
     private final Exchange exchange;
 
     private final BindingDao bindingDao;
 
     private final Map<String, BindingSet> unmodifiableBindingSetView;
 
-    public BindingsRegistry(Exchange exchange, BindingDao bindingDao) {
-        this.routingKeyToBindingMap = new HashMap<>();
+    private final BindingDeleteListener bindingDeleteListener;
+
+    BindingsRegistry(Exchange exchange, BindingDao bindingDao) {
+        this.bindingPatternToBindingsMap = new HashMap<>();
         this.exchange = exchange;
         this.bindingDao = bindingDao;
-        this.unmodifiableBindingSetView = Collections.unmodifiableMap(routingKeyToBindingMap);
+        this.unmodifiableBindingSetView = Collections.unmodifiableMap(bindingPatternToBindingsMap);
+        bindingDeleteListener = new BindingDeleteListener();
     }
 
-    void bind(Queue queue, String bindingKey, FieldTable arguments) throws BrokerException {
-        BindingSet bindingSet = routingKeyToBindingMap.computeIfAbsent(bindingKey, k -> new BindingSet());
+    void bind(QueueHandler queueHandler, String bindingKey, FieldTable arguments) throws BrokerException,
+                                                                                         ValidationException {
+        BindingSet bindingSet = bindingPatternToBindingsMap.computeIfAbsent(bindingKey, k -> new BindingSet());
+        Queue queue = queueHandler.getQueue();
         Binding binding = new Binding(queue, bindingKey, arguments);
         boolean success = bindingSet.add(binding);
-        if (success && queue.isDurable()) {
-            bindingDao.persist(exchange.getName(), binding);
+
+        if (success) {
+            queueHandler.addBinding(binding, bindingDeleteListener);
+            if (queue.isDurable()) {
+                bindingDao.persist(exchange.getName(), binding);
+            }
         }
     }
 
     void unbind(Queue queue, String routingKey) throws BrokerException {
-        BindingSet bindingSet = routingKeyToBindingMap.get(routingKey);
+        BindingSet bindingSet = bindingPatternToBindingsMap.get(routingKey);
         if (queue.isDurable()) {
             bindingDao.delete(queue.getName(), routingKey, exchange.getName());
         }
         bindingSet.remove(queue);
 
         if (bindingSet.isEmpty()) {
-            routingKeyToBindingMap.remove(routingKey);
+            bindingPatternToBindingsMap.remove(routingKey);
```

Comment:
Shouldn't we remove the listener we registered in queue handler when unbinding?

Reply:
Fixed!